Improve caching of wc_count_comments counts to avoid re-queries after every order note.

prettyboymp opened the pull request:
### Changes proposed in this Pull Request:

The following changes are made in this PR:

1. Switches the comment count caching from the transient interface to wp_cache. This allows the wp_cache_incr/decr methods to be used and is more aligned with only using transients for much slower actions.
2. Updates the callback handling for hooks related to comment modification to increment/decrement the cache wc_count_comments values for comment and post types that aren't included in the counted results stored in cache.
3. Updates the caching callback from `wp_count_comments` to use `get_comments()` and the `comments_clauses` filters rather than a custom query. So other extensions can also filter the counts as needed.

Closes #54286.  

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Setup:
1. A site running persistent object cache is required to see these improvements.
2. Use a plugin like Query Monitor to view queries.
3. Optionally add a call to `wp_cache_set_comments_last_changed();` in the theme, mu-plugins, etc where it will mimic the default cache busting for WP_Comments_Query when a new comment is created.  This just makes sure any caching issues aren't being covered up by the built in cache.

Note: You will come across a query like `SELECT COUNT(*) FROM wp_comments .. AND comment_type IN ('review', '', 'comment')`.  This is from the Product Reviews pending count bubble in the admin menu.  It is outside the scope of this PR and additional caching will need to be added for it in another PR - ref: #55015. 

#### Admin Menu Comments Bubble

1. Go to an admin page that doesn't include comment counts on it's own, e.g. like Admin -> Tools.
2. On the first visit, you may see queries like the below to count the comments by status, however, any repeat requests should not contain the queries as they will have already been cached
  ```
SELECT COUNT(*)
FROM wp_comments
LEFT JOIN wp_posts AS wp_posts_to_exclude_reviews
ON comment_post_ID = wp_posts_to_exclude_reviews.ID
WHERE ( comment_approved = '1' )
AND comment_type != 'order_note'
AND comment_type != 'webhook_delivery'
AND comment_type != 'action_log'
AND wp_posts_to_exclude_reviews.post_type NOT IN ('product')
  ```
3. Visit a post on the site and submit a comment.
4. Refresh the admin page and verify that the above query did not need to run again.
5. Open an order to edit and add and Order Note.
6. Again, refresh the admin page and verify that the above query did not need to run again.

#### Comments Admin

1. Visit the Comments Admin page.
2. Check that count queries are not needed to render the comment status filters at the top of the page.
3. Visit each comment status and verify that the item count for each page matches the count in the status filter menu.
4. Update the status of some comments, refresh the page, and verify that the counts are updated properly.
5. Go to a post on the front end and post a comment.  Verify that the comment count is updated.
6. Go to a product on the front end and post a review.  Verify that the review isn't filtered after a refresh on the product and that the comment count was not changed by the review.

Note: The item count next to the pagination is not currently cached.  This is outside the scope of this PR and may be addressed in the future.

#### Reviews Admin

1. Visit the Reviews Admin
2. Verify that the counts and reviews are displaying the correct numbers and are not modified by the additional filters.

prettyboymp committed:
Limit clearing of wc_count_comments transient to applicable comments.

Updates the callback handling for hooks related to comment modification to only clear the wc_count_comments transient for comment and post types that aren't included in the counted results stored in cache.

prettyboymp committed:
lint fixes

joshuatf commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -215,6 +215,48 @@ public static function clear_transients( $post_id ) {
+	 */
+	public static function delete_comments_count_cache_on_wp_insert_comment( $comment_id, $comment ) {
+		if ( ! self::is_comment_excluded_from_wp_comment_counts( $comment ) ) {
+			self::delete_comments_count_cache();
```
Is it possible to adjust the cache count on insert if the comment is qualified for the count (similar to https://github.com/woocommerce/woocommerce/pull/54034)?  Or do we have to remove the cache and recalculate this on these occassions?

joshuatf commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -215,6 +215,48 @@ public static function clear_transients( $post_id ) {
+	 */
+	public static function delete_comments_count_cache_on_comment_status_change( $new_status, $old_status, $comment ) {
+		if ( ! self::is_comment_excluded_from_wp_comment_counts( $comment ) ) {
+			self::delete_comments_count_cache();
```
Same question as above.

joshuatf reviewed (commented):
Thanks for tackling this, Michael!  Looks good overall; was just curious if we were able to bump the cache numbers without deleting the cache.

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
switch to incrementing/decrementing the count instead of deleting the cache.

prettyboymp commented:
I updated to increment/decrement the counts instead of clearing the cache.  We'll want to implement a job to occasionally refresh the cache to avoid any drift that may occur on systems that don't have an atomic way to increment/decrement.

prettyboymp committed:
fix update callback name to reflect 9.9.0

prettyboymp committed:
Automatically expire cache after 3 days until we can implement a scheduled job to reprime the data.

prettyboymp commented:
@joshuatf , I switched this over to use incr/decr methods to avoid having to clear the cached counts.  I also went ahead and set a 3 Day timeout for the cache as a place holder to account for any potential drift until we start using jobs to regenerate the cached counts.

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

joshuatf commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -17,6 +17,11 @@
+	/**
+	 * @var string COMMENT_COUNT_CACHE_GROUP The cache group to use for comment counts.
+	 */
+	private CONST COMMENT_COUNT_CACHE_GROUP = 'wc_comment_counts';
```
```suggestion
	private const COMMENT_COUNT_CACHE_GROUP = 'wc_comment_counts';
```

joshuatf commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -45,8 +50,8 @@ public static function init() {
 		add_filter( 'wp_count_comments', array( __CLASS__, 'wp_count_comments' ), 10, 2 );
```
It took me a really long time to debug an issue I was having here that turned out to be Docket Cache.  [It also attempts to cache `wp_count_comment` stats](https://github.com/nawawi/docket-cache/blob/0d4482a099508c696b7c5f2314b3077cec6710d8/includes/src/Tweaks.php#L133-L179) and in doing so, overrides ours.

A quick fix is to use `PHP_INT_MAX` for the priority, but this only works because (usually) the `woocommerce/` plugin folder will alphabetically run after `docket-cache` since they have the same priority.

My initial thinking was that Docket Cache's approach was wrong, but after giving this more thought, our approach with caching comments and our query below only works if there are not other comment types or exclusionary rules applied to `wp_count_comments`, otherwise it's just a matter of whoever filters `wp_count_comments` last.

To allow other plugins to continue to extend this filter, ideally we would check if `$counts` passed to the `wp_count_comments` filter was not empty and subtract our counts that we would like to exclude as opposed to directly returning a set of included counts.  For example:

```php
$counts (
    [approved] => 3
    [moderated] => 0
    [spam] => 0
    [trash] => 0
    [post-trashed] => 0
    [all] => 3
    [total_comments] => 3
)
$product_counts (
    [approved] => 1
    [moderated] => 0
    [spam] => 0
    [trash] => 0
    [post-trashed] => 0
    [all] => 2
    [total_comments] => 2
)
$diff = array_intersect( $counts, array_diff( $counts, $product_counts ) );
return $diff;
```

However, the initial consumer of the filter would need to use something like `get_comment_count( 0 )` to populate the initial value.  Another option is just to filter the comment query so that counts are not included unless the comment type is specified:

```php
add_action( 'pre_get_comments', function( $query ) {
    if ( ! isset( $query->query_vars['type'] ) ) {
        $not_in = (array) $query->query_vars['type__not_in'];
        $not_in[] = 'review';
        $query->query_vars['type__not_in'] = $not_in;
    }
} );
```

But it seems like our existing logic has been around for some time and I couldn't find any issues related to this so maybe a quick patch related to Docket Cache is enough and the comment type count conflict is too much of an edge case.

prettyboymp replied:
Good catch on this. I addressed this in [0d40f40](https://github.com/woocommerce/woocommerce/pull/54984/commits/0d40f402ecfe45ae734b8dfa40ffcad1ae9c1d00).  Responding to some of this in the top level comments.

joshuatf reviewed (commented):
This is testing really well overall and is really great to cut down on the number of times we see this query.

I ran into a rather tricky issue with Docket Cache that clashes with our own comments cache and went down a bit of a rabbit hole.  I left some comments around this, but it may be that a simple fix is in order.  I do want to at least get that plugin compatible since it's relatively popular outside of Redis and Memcached for object persistence.

Also, the testing instructions I think may be from before the cache increment updates, but here's what I tested and seemed to work well (please let me know if any of this is incorrect):

* ✅ Adding logging for cache hit/miss and cache is consistently hit after the first query.
* ✅ Leaving a comment increments the comment count in the dashboard without the need to re-query
* ✅ Adding order notes does not affect the comment count cache
* ✅ Adding product reviews does not affect the comment count cache
* ⚠️ Comment count is incorrectly overridden when Docket Cache is enabled

(Images are from Docket Cache enabled, note the missing pending comments despite a count existing)
<img width="174" alt="Screenshot 2025-03-04 at 11 23 41 AM" src="https://github.com/user-attachments/assets/3b35488c-50d8-4ac2-b03a-5b4e28fcced3" />
<img width="658" alt="Screenshot 2025-03-04 at 11 28 15 AM" src="https://github.com/user-attachments/assets/b586da52-8de3-4775-8cde-9522700adad2" />

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
Switch to using get_comments() instead of custom query so any other extensions can continue to modify the comment query as needed

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
Modify tests for non-screen based filtering of product reviews

prettyboymp committed:
restore accidentally deleted dataprovider

prettyboymp commented:
@joshuatf, I've made some adjustments based on your feedback.

I've updated the testing instructions to better reflect what this PR has turned into.

> It took me a really long time to debug an issue I was having here that turned out to be Docket Cache. [It also attempts to cache wp_count_comment stats](https://github.com/nawawi/docket-cache/blob/0d4482a099508c696b7c5f2314b3077cec6710d8/includes/src/Tweaks.php#L133-L179) and in doing so, overrides ours.
> 
> A quick fix is to use PHP_INT_MAX for the priority, but this only works because (usually) the woocommerce/ plugin folder will alphabetically run after docket-cache since they have the same priority.
> 
> My initial thinking was that Docket Cache's approach was wrong, but after giving this more thought, our approach with caching comments and our query below only works if there are not other comment types or exclusionary rules applied to wp_count_comments, otherwise it's just a matter of whoever filters wp_count_comments last.

I do think Docket Cache's approach is slightly wrong.  It is using a priority of `PHP_INT_MAX` and ignores any previously filtered `$stats` it receives.  It should probably only do one of these as this approach essentially means no other plugin can filter that hook.

> To allow other plugins to continue to extend this filter, ideally we would check if $counts passed to the wp_count_comments filter was not empty and subtract our counts that we would like to exclude as opposed to directly returning a set of included counts.

I looked into this, but didn't feel that we could confidently know the content of the `$stats` being passed in had already filtered some of the comments we were attempting to filter out and cause the numbers to be off.  My solution was to stop using a custom query, and switch to using the standard `get_comments()` like WordPress core already uses.  I then added the needed adjustments to the current `comments_clause` filters to make these properly filter out the comments we didn't want to be counted/shown outside of specific contexts. We were already running some of these filters, which is probably why the issue hasn't come up before with counts being off for a plugin like Docket Cache.

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

laurendavissmith commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -215,80 +240,130 @@ public static function clear_transients( $post_id ) {
+			$count     = wp_cache_get( $cache_key, self::COMMENT_COUNT_CACHE_GROUP );
+			if ( false === $count ) {
+				$count = get_comments( array_merge( $args, array( 'status' => $query_status ) ) );
+				wp_cache_set( $cache_key, $count, self::COMMENT_COUNT_CACHE_GROUP, 3 * DAY_IN_SECONDS );
```
Trivial but 3 * DAY_IN_SECONDS should be set to a constant

laurendavissmith commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -215,80 +240,130 @@ public static function clear_transients( $post_id ) {
+			$cache_key = 'wc_count_comments_' . $comment_query_status_to_comment_status_mapping[ $query_status ];
+			$count     = wp_cache_get( $cache_key, self::COMMENT_COUNT_CACHE_GROUP );
+			if ( false === $count ) {
+				$count = get_comments( array_merge( $args, array( 'status' => $query_status ) ) );
```
Shouldn't we cast this to (int) in the event that it fails?

joshuatf commented on plugins/woocommerce/src/Internal/Admin/ProductReviews/ReviewsUtil.php:
```diff
@@ -10,17 +10,52 @@ class ReviewsUtil {
+		/**
+		 * Any comment queries with these values are likely to be custom handling where we don't want to change default behavior.
+		 */
+		foreach ( [ 'ID', 'parent', 'parent__in', 'post_author__in', 'post_author', 'post_name', 'type', 'type__in', 'type__not_in', 'post_type__in' ] as $arg ) {
```
Should we also add `comment__in` and `comment__not_in` to this list of arguments?

joshuatf commented on plugins/woocommerce/src/Internal/Admin/ProductReviews/ReviewsUtil.php:
```diff
@@ -10,17 +10,52 @@ class ReviewsUtil {
 		}
 
+		$clauses['join']  .= " LEFT JOIN {$wpdb->posts} AS wp_posts_to_exclude_reviews ON comment_post_ID = wp_posts_to_exclude_reviews.ID ";
+		$clauses['where'] .= ( $clauses['where'] ? ' AND ' : '' ) . " wp_posts_to_exclude_reviews.post_type NOT IN ('product') ";
```
I'm a little surprised we don't wrap the `$clauses['where']` with a `trim()` for safety, but this exact same logic is used in a bunch of other places across this repo and others without issue, so it may not be a likely scenario.

prettyboymp replied:
Good call. I'm adding trim() to all the usages since comments are covered in this PR to prevent any future/current cases where another plugin adding a value like `' '` causing a failed query.

joshuatf commented on plugins/woocommerce/src/Internal/Admin/ProductReviews/ReviewsUtil.php:
```diff
@@ -10,17 +10,52 @@ class ReviewsUtil {
-			$clauses['where'] .= ( $clauses['where'] ? ' AND ' : '' ) . " wp_posts_to_exclude_reviews.post_type NOT IN ('product') ";
+		if ( ! empty( $comment_query->query_vars['post__in'] ) ) {
+			$post_ids = wp_parse_id_list( $comment_query->query_vars['post__in'] );
+			_prime_post_caches( $post_ids );
```
Nice 💯

kalessil replied:
Do we want to prime parts behind two additional parameters this function has (edit: or consider partial priming)?

prettyboymp replied:
> Do we want to prime parts behind two additional parameters this function has (edit: or consider partial priming)?

Good catch on that.  We only needed the post object for these.

joshuatf reviewed (commented):
This is testing really well for me and I like the new solution!

* ✅ Admin comment counts are appearing correctly
* ✅ Admin product review counts are appearing correctly
* ✅ Counts for both post comments and product reviews are shown correctly on the frontend
* ✅ Cache is consistently hit and updated on comment status change

<img width="213" alt="Screenshot 2025-03-10 at 3 46 10 PM" src="https://github.com/user-attachments/assets/578db317-9172-4e92-8333-205ddde6e567" />
<img width="472" alt="Screenshot 2025-03-10 at 3 44 26 PM" src="https://github.com/user-attachments/assets/b37cfb24-402d-4656-9b6c-ce32afaa4ae8" />

Added one comment about included some additional arguments for `WP_Comment_Query` where we should ignore our clauses, but otherwise this LGTM.

prettyboymp committed:
excluding product type filter for comment__in/not_in as likely custom queries
checking the trim'ed state of the where clause when determining whether to prefix with 'AND'

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
lint fixes

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
more lint fixes

prettyboymp committed:
lint fixes in test

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

kalessil commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -43,8 +43,11 @@ public static function init() {
+
 		// Exclude product reviews.
-		add_filter( 'comments_clauses', array( ReviewsUtil::class, 'comments_clauses_without_product_reviews' ), 10, 1 );
+		add_filter( 'comments_clauses', array( ReviewsUtil::class, 'comments_clauses_without_product_reviews' ), 10, 2 );
```
Don't have full context here, but it seems that `comment_feed_where` handling was not added for the two new handlers of `comments_clauses`. If it's intentional, can we leave some comments on why we skipped them?

kalessil commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -230,7 +250,10 @@ public static function clear_transients( $post_id ) {
 	public static function increment_comments_count_cache_on_wp_insert_comment( $comment_id, $comment ) {
 		if ( ! self::is_comment_excluded_from_wp_comment_counts( $comment ) ) {
-			wp_cache_incr( 'wc_count_comments_' . $comment->comment_approved, 1, self::COMMENT_COUNT_CACHE_GROUP );
+			$comment_status = wp_get_comment_status( $comment );
```
`$comment_status = wp_get_comment_status( $comment );` good one!

kalessil commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -245,14 +268,7 @@ public static function increment_comments_count_cache_on_wp_insert_comment( $com
-			$old_status     = $old_status_map[ $old_status ] ?? $old_status;
-
-			wp_cache_incr( 'wc_count_comments_' . $comment->comment_approved, 1, self::COMMENT_COUNT_CACHE_GROUP );
+			wp_cache_incr( 'wc_count_comments_' . $new_status, 1, self::COMMENT_COUNT_CACHE_GROUP );
```
Nit:

```php
    $is_incremented = false !== wp_cache_incr( 'wc_count_comments_' . $new_status, 1, self::COMMENT_COUNT_CACHE_GROUP );
    if ( $is_incremented ) {
        wp_cache_decr( 'wc_count_comments_' . $old_status, 1, self::COMMENT_COUNT_CACHE_GROUP );
    }
```

prettyboymp replied:
The two cached counts aren't reliant on each other.  `wp_cache_incr( 'wc_count_comments_' . $new_status );` will return `false` if it has fallen out of cache and regenerate the next request to retrieve it.
We still need to attempt to decrement `'wc_count_comments_'. $old_status`, otherwise, if it is still cached, it will be off by one since it won't regenerate.

kalessil replied:
:+1:

kalessil commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -280,76 +296,74 @@ public static function delete_comments_count_cache() {
 			);
-			foreach ( $comment_approved_values as $approved_value ) {
-				wp_cache_delete( 'wc_count_comments_' . $approved_value, self::COMMENT_COUNT_CACHE_GROUP );
+			foreach ( $comment_statuses as $comment_status ) {
```
Nit: `wp_cache_delete_multiple`

kalessil commented on plugins/woocommerce/includes/class-wc-comments.php:
```diff
@@ -215,80 +242,130 @@ public static function clear_transients( $post_id ) {
 	public static function delete_comments_count_cache() {
-		delete_transient( 'wc_count_comments' );
+		if ( wp_cache_supports( 'flush_group' ) ) {
+			wp_cache_flush_group( self::COMMENT_COUNT_CACHE_GROUP );
```
I'm also using this group in https://github.com/woocommerce/woocommerce/pull/56215/files#diff-389d36391ae130391874a998ea6e4a3ab4a6b902f223e6e7c1ba0c464579b03dR18 and I think we'll need to avoid `wp_cache_flush_group( self::COMMENT_COUNT_CACHE_GROUP )`.

Given my other comment prompting `wp_cache_delete_multiple` usage we should be good with just using it.

prettyboymp committed:
Merge branch 'trunk' into fix/54286-limit-wc-comment-count-transient-clearing

prettyboymp committed:
excluded action log entries from comment feed
avoided flushing cache group to directly delete specific keys for status

prettyboymp commented:
@kalessil updates are in [95d1af2](https://github.com/woocommerce/woocommerce/pull/54984/commits/95d1af2abc3278e85daf6265609de6841fe1ee14)

kalessil reviewed (approved):
LGTM!

:heavy_check_mark: Followed the testing instructions - works as intended
:heavy_check_mark: Tested bulk operations
:heavy_check_mark: Tested random status transitions for individual and multiple comments
:heavy_check_mark: Did product reviews smoke testing - not affected
